### ISSUE TITLE
[as2_simulation_gazebo] Does the gripper work or how to include custom plugins

### ISSUE THREAD

**Issue by mhl787156** [aerostack2/aerostack2#689]
### ROS2 Version

humble

### Version

main

### Installation

binaries

### What happened?

Suction gripper is currently not a valid payload target, although the bridges and implementation exists. 

Is this on purpose and Does it work?

If not, what is the current recommended way of including your own plugin (e.g. simple pickup plugin) and adding that to quadrotor model (create copy of model locally with plugin included? create a new model locally which references quadrotor_base which includes the new plugin? Just create a new "payload" and somehow use your system to add to the drone via json?) 

> If this, due to your strict checking of allowed models, would you be able to prepend entries from `GZ_SIM_RESOURCE_PATH` so that my custom version of e.g. quadrotor_base could be loaded first (models/drone.py)?  Or allow other models but with an *at your own risk* warning. 
 
Currently using gazebo fortress. 

![image](https://github.com/user-attachments/assets/66ef4be9-0447-4dea-a26c-814de1c52769)

```
{
    "world_name": "empty_delivery",
    "drones": [
        {
            "model_type": "quadrotor_base",
            "model_name": "cf0",
            "xyz": [
                0.0,
                0.0,
                0.3
            ],
            "rpy": [
                0,
                0,
                0.0
            ],
            "flight_time": 60,
            "payload": [
                {
                    "model_name": "hd_camera",
                    "model_type": "hd_camera",
                    "rpy": [0.0, 1.58, 0.0]
                },
                {
                    "model_name": "suction_gripper",
                    "model_type": "suction_gripper",
                    "xyz": [0.0, 0.0, 0.5]
                }
            ]
        }
    ],
...
```

### How to reproduce your issue?

I use Ubuntu 22.04, Gazebo Fortress, using aerostack2 v1.1.2

**Comment by fjanguita** [comment]
Hi @mhl787156

The SuctionGripper plugin is not compiled in the current release even though the implementation exists, that's why the gripper is not added as a valid payload.

To try out a new custom plugin, you can add it directly to your own drone model or you can create a new model and register it as a payload. Both options are valid, but you will have to manipulate Aerostack2 source code one way or the other. Please check the [Create New Gazebo Assets](https://aerostack2.github.io/_09_development/_develop_guide/_develop_guide/_create_gazebo_assets.html) guide and the [Adding New Gazebo Assets](https://aerostack2.github.io/_09_development/_tutorials/_tutorials/adding_gazebo_assets.html) tutorial from the Aerostack2 documentation.

If you end up using your own plugin, remember to set the `GZ_SIM_SYSTEM_PLUGIN_PATH` environment variable to find the directory to which your plugin `.so` file is exported.

**Comment by mhl787156** [comment]
Hi @fjanguita 

Thanks for the reply. 

Ah I see, thats a shame it would be good if it could be included in a future release! 

On adding new plugins, it would be great if there could be a way to add new assets without modifying the as2 `DroneTypeEnum`. Maybe if I could pass my own version of DroneTypeEnum into the pydantic checker - though I have no idea how that could work. For teaching especially, it means that I can just get students to pull standard as2 while running our custom model. 

For now I'll just fork a modified version instead! 

Thanks

**Comment by pariaspe** [comment]
Hi @mhl787156 

Yes, we agree on that. We have an idea to include `Enum` on world description file, but we have to carefully modify pydantic entities. However, we are a bit overwhelmed lately so I can't promise it will be included soon

**Comment by mhl787156** [comment]
Hi @pariaspe 

No worries, you guys are becoming popular! Wasn't expecting you guys to sort it immediately!

I'm not familiar with Pydantic, do they have some sort of wild card matching which we could point to a BaseVehicleType to use in the interim?

**Comment by pariaspe** [comment]
> I'm not familiar with Pydantic, do they have some sort of wild card matching which we could point to a BaseVehicleType to use in the interim?

The drone dataclass currently in use is:
```python
class Drone(Entity):
    model_name: str
    model_type: DroneTypeEnum
    xyz: conlist(float, min_items=3, max_items=3) = [0, 0, 0]
    rpy: conlist(float, min_items=3, max_items=3) = [0, 0, 0]
    flight_time: int = 0  # in minutes
    battery_capacity: float = 0  # Ah
    payload: List[Payload] = []
    enable_velocity_control: bool = True
```
Notice that only `model_name` and `model_type` are required. Any other fields are optional. `battery_capacity` is calculated on construction (pydantic calls it "validators") based on the `flight_time` given. 

My idea, which is open to discuss, is to also accept strings as model types: `model_type: DroneTypeEnum | str`. The problem here is that any string would be valid, so pydantic validation won't prevent errors in model types anymore. Anyway, if model is not found, we are raising an exception telling the user that the model is not found.

https://github.com/aerostack2/aerostack2/blob/7be5f6a3bcbcb128ed33f9f056e31a3690cc1808/as2_simulation_assets/as2_gazebo_assets/src/as2_gazebo_assets/models/drone.py#L176-L184

An alternative would be to check (in a pydantic validator) if the new model type introduced via string is in a Environment Variable (that the user should configure) and if not discard it during the validation. I'm not fully convinced with this, seems a bit redundant to me.

**Comment by mhl787156** [comment]
Perhaps a combination of the two might work? Since I'm assuming custom models are "advanced behaviour" could you add an boolean environment variable "AS2_ENABLE_CUSTOM_VEHICLE". If enabled, that allows a string input for `model_type` but you've explicitly made it the developers problem. If disabled then it is as now.

Other than that, some sort of inheritance structure for extending the enum definition? No idea if pydantic can check on that. 

For now I will just add my drone to my fork of as2 :) Thanks for all the help!

**Comment by fjanguita** [comment]
Hi @mhl787156,

The issue we were discussing was solved by PR #699 mentioned above. If you try to add a new drone model as a model_type, Aerostack2 will now look for a model folder with the same name as the model_type you specified in one of the added Gazebo resources directories.

The issue will be closed but feel free to open it again if you need anything else.

Thank you for your comments. I hope you find this helpful :)

**Comment by luccagandra** [comment]
I managed to compile the SuctionGripper plugin for Gazebo Fortress. It seems that the plugin version is older compared to the current one in the [mbzirc](https://github.com/osrf/mbzirc/blob/main/mbzirc_ign/src/SuctionGripper.cc) repository

I'm working on making it work with Harmonic as well.
